### DISCUSSION TITLE
TimeStream Query

### DISCUSSION THREAD

**Discussion by mjones-asasolutions** [awslabs/aws-sdk-swift#1417] [Q&A] [answered]
Hi all,

I don't know if anyone else has experienced this, but I'm not able to pull back any data using TimeStreamQuery using this SDK.  I have no issues when using the CLI or Postman to retrieve this same TimeStream query.  Also, the SDK works fine for me for all of the other libraries and calls I've been making with it.

Below is a copy of the code I'm using to test

```
let qStringTest: String = "\"SELECT * FROM \'ConnectRealTime\'.\'Queue\' WHERE time between ago(15d) and now() ORDER BY time DESC LIMIT 10\""

func timestreamTest() async throws -> [String] {
    print("Step 1 - A")
    let client = try TimestreamQueryClient(region: "us-east-1")
    print(client)
    print("Step 1 - B")
    print(qStringTest)
    let output = try await client.query(
        input: QueryInput(
            queryString: qStringTest
        )
    )
    print("Step 2")
    // Process the output to extract necessary information
    var resultStrings: [String] = []
    for row in output.rows ?? [] {
        print("Step 2 - row")
        for data in row.data ?? [] {
            // Check if scalar value exists and is convertible to a string
            if let scalarValue = data.scalarValue {
                // Convert scalar value to a string and add to resultStrings
                let stringValue = String(describing: scalarValue)
                resultStrings.append(stringValue)
            }
        }
    }
    print("Step 3")
    print("TEST START: \(resultStrings)")
    print("***TEST END***")
    // Return the processed strings
    return resultStrings
}
```

In the console, I'm getting a I'm getting `UnknownAWSHTTPServiceError` and a 404 status code. 
```
2024-03-21T18:32:14-0400 info URLSessionHTTPClient : [Logging] urlSession(_:dataTask:didReceive:) called
2024-03-21T18:32:14-0400 info URLSessionHTTPClient : [Logging] urlSession(_:dataTask:didReceive:) called with 63 bytes
2024-03-21T18:32:14-0400 info URLSessionHTTPClient : [Logging] urlSession(_:task:didCompleteWithError:) called. Success
Error: UnknownAWSHTTPServiceError(typeName: Optional("UnknownOperationException"), message: nil, requestID: nil, requestID2: nil, httpResponse: 
Status Code: http_status_404 
 Date: Thu, 21 Mar 2024 22:32:13 GMT, 
Content-Length: 63, 
Content-Type: application/x-amz-json-1.0, 
x-amzn-RequestId: c6539b22-3e3b-4c37-8779-544cb38a7332)
```

Any help or advice would be greatly appreciated.
Thank you!!!

**Comment by dayaffe** [comment]
Thank you for your question! We will investigate and get back to you here when we have more info.

**Comment by sichanyoo** [comment, marked as answer]
The TimestreamQuery service requires that the API endpoint for queries be discovered using the DescribeEndpoints action. See the docs for DescribeEndpoints [here](https://docs.aws.amazon.com/timestream/latest/developerguide/API_query_DescribeEndpoints.html).

Because the AWS SDK for Swift is still in developer preview and not yet feature complete, automatic endpoint discovery is not yet implemented. Eventually endpoint discovery will be performed transparently by the SDK when needed, but as a temporary workaround you can call DescribeEndpoints directly before querying, then use the returned endpoint to create a client with the right endpoint for querying.

Here is sample code that will retrieve records from a Timestream sample database:

    do {
        let endpointDiscoveryClient = try TimestreamQueryClient(region: "us-east-1")
        let description = try await endpointDiscoveryClient.describeEndpoints(input: DescribeEndpointsInput())
        guard let hostname = description.endpoints?.first?.address else {
            fatalError("Hostname not provided in description")
        }
        let config = try TimestreamQueryClient.TimestreamQueryClientConfiguration(endpoint: "https://\(hostname)", region: "us-east-1")
        let client = TimestreamQueryClient(config: config)
        let input = QueryInput(queryString: "SELECT * FROM sampleDB.IoTMulti LIMIT 10")
        let output = try await client.query(input: input)
        print("Request succeeded")
    } catch {
        print("Request failed")
        print("Error: \(error)")
    }
Please try this approach and see if this fixes your issue.

**Reply by mjones-asasolutions** [reply]
Thank you!!!  

I'll give this a try

**Reply by sichanyoo** [reply]
Hello, just checking in. Did that resolve the issue?

**Reply by mjones-asasolutions** [reply]
Hi,

I’m actually travelling this week, but I intend to try it as soon as I get back next week.

**Reply by sichanyoo** [reply]
Gotcha, thank you for the heads up!

**Reply by mjones-asasolutions** [reply]
It worked!  Thank you so much!
